### PR TITLE
Plastic BRDF

### PR DESCRIPTION
**Pull request by est77**
Similar to Mitsuba's rough plastic.

plastic_tester

### CODE REVIEW

File: src/appleseed/renderer/modeling/bsdf/plasticbrdf.cpp [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+    // References:
+    //
+    //   [1] A Physically-Based Reflectance Model Combining Reflection and Diffraction
+    //       https://hal.inria.fr/hal-01386157
+    //
+    //   [2] Mitsuba renderer
+    //       https://github.com/mitsuba-renderer/mitsuba
+    //
+
+    const char* Model = "plastic_bsdf";
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/plasticbrdf.cpp]: That should probably be `plastic_brdf`.

File: src/appleseed/renderer/modeling/bsdf/plasticbrdf.cpp [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+            SamplingContext&        sampling_context,
+            const void*             data,
+            const bool              adjoint,
+            const bool              cosine_mult,
+            BSDFSample&             sample) const APPLESEED_OVERRIDE
+        {
+            const Basis3f& shading_basis(sample.m_shading_basis);
+            const Vector3f& n = shading_basis.get_normal();
+            const Vector3f& outgoing = sample.m_outgoing.get_value();
+            const float cos_on = std::min(dot(outgoing, n), 1.0f);
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/plasticbrdf.cpp]: Remove redundant namespace qualifier.

File: src/appleseed/renderer/modeling/bsdf/plasticbrdf.cpp [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+            if (cos_on < 0.0f)
+                return;
+
+            const InputValues* values = reinterpret_cast<const InputValues*>(data);
+            const float alpha = microfacet_alpha_from_roughness(values->m_roughness);
+
+            // Compute the microfacet normal by sampling the MDF.
+            Vector3f wo = shading_basis.transform_to_local(outgoing);
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/plasticbrdf.cpp]: Make const.

File: src/appleseed/renderer/modeling/bsdf/plasticbrdf.cpp [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+            sampling_context.split_in_place(4, 1);
+            const Vector4f s = sampling_context.next2<Vector4f>();
+            Vector3f m = m_mdf->sample(wo, Vector3f(s[0], s[1], s[2]), alpha, alpha);
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/plasticbrdf.cpp]: Make const.

File: src/appleseed/renderer/modeling/bsdf/plasticbrdf.cpp [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+            const Vector3f wo = shading_basis.transform_to_local(outgoing);
+            const Vector3f wi = shading_basis.transform_to_local(incoming);
+
+            const Vector3f m = normalize(wi + wo);
+            const float Fo = fresnel_term(values->m_precomputed.m_eta, wo, m);
+            const float Fi = fresnel_term(values->m_precomputed.m_eta, wi, m);
+
+            const float alpha = microfacet_alpha_from_roughness(values->m_roughness);
+
+            float probability = 0.0;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/plasticbrdf.cpp]: `0.0f`

File: src/appleseed/renderer/modeling/bsdf/plasticbrdf.cpp [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+
+            const Vector3f wo = shading_basis.transform_to_local(outgoing);
+            const Vector3f wi = shading_basis.transform_to_local(incoming);
+            const Vector3f m = normalize(wi + wo);
+
+            const float Fo = fresnel_term(values->m_precomputed.m_eta, wo, m);
+
+            float probability = 0.0f;
+
+            if (ScatteringMode::has_glossy(modes))
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/plasticbrdf.cpp]: Add curly braces around multi-line blocks.

File: src/appleseed/renderer/modeling/bsdf/plasticbrdf.cpp [hunk 7]
```diff
@@ -0,0 +1,554 @@
...
+      private:
+        typedef PlasticBRDFInputValues InputValues;
+
+        static float fresnel_term(
+            const float&            eta,
+            Vector3f                wo,
+            const Vector3f&         m)
+        {
+            float f;
+            foundation::fresnel_reflectance_dielectric(
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/plasticbrdf.cpp]: Remove redundant namespace qualifiers.

**Comment by dictoon** [comment]
Looking pretty good! Just a few minor things to fix and it's good to merge.

I like the fact that the internal scattering parameter is a float, not a boolean.

**Comment by dictoon** [comment]
Make sure to update the test scenes as well.

**Comment by est77** [comment]
Should be good to merge now.

**Comment by dictoon** [comment]
Thank you!

**Pull request merged**